controllers/progressivesync_controller.go

```diff
@@ -418,12 +431,44 @@ func (r *ProgressiveSyncReconciler) syncApp(ctx context.Context, appName string)
 	return r.ArgoCDAppClient.Sync(ctx, &syncReq)
 }
 
-// patchStatus patches the progressive sync object status
-func (r *ProgressiveSyncReconciler) patchStatus(ctx context.Context, pr *syncv1alpha1.ProgressiveSync) error {
-	key := client.ObjectKeyFromObject(pr)
-	latest := syncv1alpha1.ProgressiveSync{}
-	if err := r.Client.Get(ctx, key, &latest); err != nil {
-		return err
-	}
-	return r.Client.Status().Patch(ctx, pr, client.MergeFrom(&latest))
+// setStatus updates the progressive sync object status with backoff
+func (r *ProgressiveSyncReconciler) setStatus(ctx context.Context, pr *syncv1alpha1.ProgressiveSync) error {
+
+	retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+
+		key := client.ObjectKeyFromObject(pr)
+		latest := syncv1alpha1.ProgressiveSync{}
+		if err := r.Client.Get(ctx, key, &latest); err != nil {
+			return err
+		}
+		latest.Status = pr.Status
+		if err := r.Client.Status().Update(ctx, &latest); err != nil {
+			return err
+		}
+		pr = &latest
+		return nil
+
+	})
+	return retryErr
+}
+
+// initializeStatus initializes the progressive sync object status to prevent conflicts when updating status later on
+func (r *ProgressiveSyncReconciler) initializeStatus(ctx context.Context, pr *syncv1alpha1.ProgressiveSync) error {
+
+	retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
```

Comment:
Why do we need the retry policy here? I thought that getting the latest object first was enough to avoid conflicts.

Reply:
We probably don't need it anymore, initially I had this method setting the state on the server side as well, we no longer do that since the reconcilation loop will eventually in all it's code paths prior to requeueing push the state to the server.

I'll see if I can remove this, nice catch! 👍

Reply:
This was removed in a0f7606772c9e9a6482961ed62463a238f4a3f71, it was not needed!

Reply:
In the end we decided to leave the retry logic and do a full status update. This is the recommended approach suggested in https://github.com/kubernetes/client-go/blob/master/examples/create-update-delete-deployment/main.go#L125